File: app/src/main/java/com/battlelancer/seriesguide/widgets/WatchedBox.java
package com.battlelancer.seriesguide.widgets;

import android.content.Context;
import android.util.AttributeSet;
import androidx.appcompat.widget.AppCompatImageView;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.ui.episodes.EpisodeFlags;
import com.battlelancer.seriesguide.ui.episodes.EpisodeTools;

/**
 * Image view that displays a watched, skipped or watch icon depending on the given episode flag.
 */
public class WatchedBox extends AppCompatImageView {

    private int episodeFlag;

    public WatchedBox(Context context, AttributeSet attrs) {
        super(context, attrs);

        if (isInEditMode()) {
            episodeFlag = EpisodeFlags.UNWATCHED;
            updateStateImage();
        }
    }

    public int getEpisodeFlag() {
        return episodeFlag;
    }

    /**
     * Set an {@link EpisodeFlags} flag.
     */
    public void setEpisodeFlag(int episodeFlag) {
        EpisodeTools.validateFlags(episodeFlag);
        this.episodeFlag = episodeFlag;
        updateStateImage();
    }

    private void updateStateImage() {
        switch (episodeFlag) {
            case EpisodeFlags.WATCHED: {
                setImageResource(R.drawable.ic_watched_24dp);
                break;
            }
            case EpisodeFlags.SKIPPED: {
                setImageResource(R.drawable.ic_skipped_24dp);
                break;
            }
            case EpisodeFlags.UNWATCHED:
            default: {
                setImageResource(R.drawable.ic_watch_black_24dp);
                break;
            }
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/episodes/WatchedBox.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2011-2024 Uwe Trottmann

package com.battlelancer.seriesguide.shows.episodes

import android.content.Context
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.TooltipCompat
import com.battlelancer.seriesguide.R

/**
 * Image view that displays a watched, skipped or watch icon depending on the given episode flag.
 *
 * Provides a content description and tooltip out of the box.
 */
class WatchedBox(context: Context, attrs: AttributeSet?) : AppCompatImageView(context, attrs) {

    /**
     * An [EpisodeFlags] flag.
     */
    var episodeFlag = 0
        set(value) {
            EpisodeTools.validateFlags(value)
            field = value
            updateStateImage()
            updateContentDescription()
        }

    init {
        if (isInEditMode) {
            episodeFlag = EpisodeFlags.UNWATCHED
        }
    }

    private fun updateStateImage() {
        when (episodeFlag) {
            EpisodeFlags.WATCHED -> {
                setImageResource(R.drawable.ic_watched_24dp)
            }

            EpisodeFlags.SKIPPED -> {
                setImageResource(R.drawable.ic_skipped_24dp)
            }

            EpisodeFlags.UNWATCHED -> {
                setImageResource(R.drawable.ic_watch_black_24dp)
            }

            else -> {
                setImageResource(R.drawable.ic_watch_black_24dp)
            }
        }
    }

    private fun updateContentDescription() {
        val watched = EpisodeTools.isWatched(episodeFlag)
        contentDescription =
            context.getString(if (watched) R.string.action_unwatched else R.string.action_watched)
        // Re-set tooltip text after updating
        TooltipCompat.setTooltipText(this, contentDescription)
    }
}
